---
title: "20240417_de_report"
author: "Jonathan Arditi"
date: "2024-04-17"
output:
  html_document:
    df_print: paged
  pdf_document: default
  word_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

##Warmup & Quality Assessment

Keep in mind the correspondence between sample names and their cell type. The big numbers are recurrent cells:

+---------+-----------+
| Primary | Recurrent |
+=========+===========+
| X12     | X115      |
|         |           |
| X153    | X79       |
|         |           |
| GC2     |           |
+---------+-----------+

```{r pt1, include=FALSE}
setwd("C:/Users/jardi/Brown Dropbox/Jonathan Arditi/Documents/RwdLocal/chordoma/in-house-rna")
source("../../res/mypca.R")
source("../../res/mor_norm.R")
source("../../res/write_ontologies.R")
library(stringr)
library(DESeq2)
library(tidyverse)
library(ggplot2)
library(pheatmap)
library(cluster)
library(biomaRt)
library(gridExtra)
library(apeglm)
library(EnhancedVolcano)
#library(dataVisEasy)
#de<-read.table("chordoma/rna-azenta/rna-counts-txid.txt", header = TRUE)
mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
r1<-read.table("115.counts.txt", header=T)
p1<-read.table("12.counts.txt", header=T)
r2<-read.table("79.counts.txt", header=T)
p2<-read.table("153.counts.txt", header=T)
p3<-read.table("GC2.counts.txt", header=T)
samplenames<-c("X115","X12","X79","X153","GC2")
#samplelabels<-factor(c("Recurrent", "Primary", "Primary", "Recurrent", "Primary"))
samplelabels<-c("Recurrent", "Primary", "Recurrent", "Primary", "Primary")
samplekey<-samplelabels
names(samplekey)<-samplenames
de<-cbind(r1, p1$X12, r2$X79, p2$X153, p3$GC2)
names(de)[7:11]<-samplenames

counts<-de[7:11]
rownames(counts)<-de$Geneid

```

```{r, echo=F}
hist(as.matrix(log10(counts)))
```

```{r, echo=FALSE}
summary(as.matrix(counts))
```

We see that most genes have very few counts but some have a great many counts.

```{r, echo=F}
counts %>%reshape2::melt() %>% ggplot(., aes(x=variable, y=log10(value))) + geom_violin()
```

Relative to the other samples, X153 seems enriched for genes with low counts, though it also has some crazy outliers.

```{r, include=F}
dds <- DESeqDataSetFromMatrix(countData = counts, 
                              colData = DataFrame(samplelabels), 
                              design = ~ samplelabels)

p_mean_sd_scaled <- 
  counts %>% 
  as.data.frame() %>% 
  rownames_to_column("gene") %>% 
  pivot_longer(cols = - gene, names_to = "sample", values_to = "counts") %>% 
  group_by(gene) %>% 
  summarise(gene_average = mean(counts), gene_stdev = sd(counts)) %>% 
  ungroup() %>% 
  ggplot(., aes(x = log10(gene_average), y = log10(gene_stdev))) +
  geom_point(alpha = 0.5, fill = "grey", colour = "black") +
  labs(x = "Gene count average (log10 scale)",
       y = "Gene count standard deviation (log10 scale)") +
  ggtitle("Mean - Standard deviation relationship\n(no variance stabilisation ")
```

```{r, echo=F}
p_mean_sd_scaled
```

We see that, as is typical in RNA data, the variance of the counts is a fucntion of the average counts. this is called heteroscedasticity. We can adjust this to make the mean and sd more independent:

```{r, include=F}
dds = estimateSizeFactors(dds)

dds = estimateDispersions(object = dds, 
                          fitType = "parametric", 
                          quiet = TRUE)

vsd = varianceStabilizingTransformation(object = dds, 
                                        blind = TRUE,           # do not take the design formula into account
                                        # best practice for sample-level QC
                                        fitType = "parametric")

# extract the matrix of variance stabilised counts
variance_stabilised_counts <- assay(vsd)

# create the mean-sd plot
p_mean_sd_vst <- 
  variance_stabilised_counts %>% 
  as.data.frame() %>% 
  rownames_to_column("gene") %>% 
  pivot_longer(cols = - gene, names_to = "sample", values_to = "counts") %>% 
  group_by(gene) %>% 
  summarise(gene_average = mean(counts), gene_stdev = sd(counts)) %>% 
  ungroup() %>% 
  ggplot(., aes(x = gene_average, y = gene_stdev)) +
  geom_point(alpha = 0.5, fill = "grey", colour = "black") +
  labs(x = "Gene count average (variance stabilised)", 
       y = "Gene count standard deviation (variance stabilised)") +
  ggtitle("Mean - Standard deviation relationship\n(after variance stabilisation ")
```

```{r, echo=F}
p_mean_sd_vst
```

```{r}
t_variance_stabilised_counts <- t(variance_stabilised_counts)
#rm(variance_stabilised_counts)
# before computing the PCA, check that samples are in rows and genes in columns
pca_results <- mypca(t_variance_stabilised_counts, 
                     center = TRUE, 
                     scale = TRUE)
#rm(t_variance_stabilised_counts)

scores <- pca_results$scores

lbls<-c("Recurrent", "Primary", "Recurrent", "Primary", "Primary")

scores_with_conditions <- 
  scores %>% 
  rownames_to_column("samplekey") %>% # to prepare to join on the "sample" column
  left_join(x = .,                 # this means that we are passing the 'scores' dataframe 
            y = data.frame(samplekey),         # this dataframe contains the sample to condition correspondence
            by = "samplekey") 

scores_with_conditions<-cbind(lbls, scores_with_conditions)
# explained variance
# one % variance value per PC
explained_variance <- 
  pca_results$explained_var %>% 
  pull("exp_var")

p<-ggplot(scores_with_conditions, 
       aes(PC1, PC2, color = lbls)) +
  geom_point(size = 4) +
  geom_text(label=samplenames, vjust=-0.5) +
  xlab(paste0("PC1: ",explained_variance[1],"% variance")) +
  ylab(paste0("PC2: ",explained_variance[2],"% variance")) + 
  coord_fixed(ratio = 1) +
  ggtitle("PCA")
```

```{r, echo=F}
plot(p)
```

```{r, echo=F}
# make the scree plot
p<-ggplot(pca_results$explained_var, 
       aes(x = seq(from = 1, to = nrow(pca_results$explained_var)), 
           y = exp_var)) +
  ylab('explained variance (%)') + 
  ggtitle('Explained variance per component') + 
  geom_bar(stat = "identity") +
  labs(x = "Principal Component number") +
  scale_x_continuous(breaks = seq(
    from = 1, 
    to = nrow(pca_results$explained_var)))
```

```{r, echo=F}
plot(p)
```

```{r, include=F}
# create a dplyr tibble
size_factors_df <- tibble(
  sample = names(sizeFactors(dds)), 
  correction_factor = sizeFactors(dds)
)

# line plot to connect the different size factor values
p <- ggplot(size_factors_df, aes(x = sample, y = correction_factor, group = 1)) +
  geom_bar(stat="identity") + 
  theme(axis.text.x = element_text(angle = 90, size = 5)) 
```

```{r, echo=F}
plot(p)
```

Although the PCs are sufficient to describe our samples (2 PCs sum to \>50% of variance) we see from all these figures that X153 is very different than the other samples in terms of counts distribution and the corrections required to bring it in line with the other samples. RNA degradation is possible.

```{r, include=F}
#once again we see that X153 (recurrent) is an outlier in terms of dispersion
counts_normalised = counts(dds, normalized = TRUE)

pt1<-counts %>%reshape2::melt() %>% ggplot(., aes(x=variable, y=log10(value))) + geom_boxplot()
pt2<-counts_normalised %>%reshape2::melt() %>% ggplot(., aes(x=Var2, y=log10(value))) + geom_boxplot()

```

```{r, echo=F}
grid.arrange(pt1, pt2, nrow=2)
```

##Differential Expression

Now we conduct our DE analysis. We see that once again we don't have great p values due to our sample n=5.

```{r, include=F}
dds <- DESeq(dds)
res <- results(dds)
all_genes_results <- results(dds, contrast = c("samplelabels",                      # name of the factor
                                               "Recurrent",    # name of the numerator level for fold change
                                               "Primary"))          
```

```{r, echo=F}
hist(all_genes_results$pvalue, col="grey", main = "Non-adjusted p-value distribution")

```

```{r, echo=F}
hist(all_genes_results$padj, col="lightblue", main = "Adjusted p-value distribution")

```

```{r, echo=F}
hist(all_genes_results$padj, col="lightblue", main = "Adjusted p-value distribution, <0.2", xlim = c(0,0.2), ylim=c(0,250), breaks=50)
```

```{r, include=F}
resLFC <- lfcShrink(dds = dds, 
                    res = all_genes_results,
                    type = "normal",
                    coef = "samplelabels_Recurrent_vs_Primary") # name or number of the coefficient (LFC) to shrink
#Note that type='apeglm' and type='ashr' have shown to have less bias than type='normal'.
#See ?lfcShrink for more details on shrinkage type, and the DESeq2 vignette.
#Reference: https://doi.org/10.1093/bioinformatics/bty895

l2<-which(abs(resLFC$log2FoldChange)>=1.5)
alldegenes<-getBM(c("hgnc_symbol", "chromosome_name", "ensembl_gene_id"), 
                  "ensembl_gene_id", 
                  values=rownames(resLFC)[l2], 
                  mart)

get_gene<-function(enssym){
  x<-alldegenes$hgnc_symbol[alldegenes$ensembl_gene_id==enssym]
  if(length(x)<1){enssym}else if (x==""){enssym}else{x}
}

newrownames<-sapply(rownames(resLFC), get_gene)
resLFC2<-resLFC
rownames(resLFC2)<-newrownames
```

```{r, echo=F}
p<-EnhancedVolcano(toptable = resLFC2,
                x = "log2FoldChange",
                y = "padj",
                lab = rownames(resLFC2),
                xlim = c(-5, +5),
                ylim = c(0,6),
                pCutoff = 5e-2,
                pointSize = 2.0,
                labSize=2.5,
                FCcutoff = 1.5, 
                title = "Recurrent versus Primary \n (fold change cutoff = 2, p-value cutoff = 1e-06)"
)
plot(p)
```

```{r, include=F}
resLFC2a<-resLFC2[(!is.na(resLFC2$log2FoldChange)) & (!is.na(resLFC2$padj)) & (abs(resLFC2$log2FoldChange)>1.5) & (resLFC2$padj<0.05),]

resLFC2a<-resLFC2a[order(resLFC2a$log2FoldChange),]

counts_normalised_de<-counts_normalised[(!is.na(resLFC2$log2FoldChange)) & (!is.na(resLFC2$padj)) & (abs(resLFC2$log2FoldChange)>1.5) & (resLFC2$padj<0.05),]
rownames(counts_normalised_de)<-sapply(rownames(counts_normalised_de), get_gene)

counts_scaled <- 
  counts_normalised_de %>% 
  t(.) %>%                              # transpose to have the genes in columns 
  scale() %>%                           # scale(x, center = TRUE, scale = TRUE) 
  t(.)                                  # back in original shape

```

```{r, echo=F}
cs2<-counts_scaled
colnames(cs2)<-c("Recurrent 1", 
                 "Primary 1", 
                 "Recurrent 2", 
                 "Primary 2", 
                 "Primary 3*")

ph<-pheatmap(cs2, 
         cluster_rows = TRUE,                      
         cluster_cols = TRUE, 
         show_rownames = TRUE, 
         show_colnames = TRUE,
         fontsize_row = 4,
         main = "Clustering by differentially expressed genes")

```

Finally I do my classic ontology analysis

```{r, echo=F}

degenes<-rownames(counts_scaled)
x<-write_ontology(degenes, "pass1_dep-r")
```
